File: 0823 - Binary Trees With Factors/solution.cpp
class Solution {
public:
    int numFactoredBinaryTrees(vector<int>& A) {
        int res = 0, MOD = 1e9+7, n = A.size();
        unordered_map<int, int> m;
        sort(A.begin(), A.end());
        for(int i = 0; i < n; i++){
            m[A[i]] = 1;
            for(int j = 0; j < i; j++){
                if(A[i]%A[j] || !m.count(A[i]/A[j])) continue;
                m[A[i]] = (m[A[i]]+(long long)m[A[j]]*m[A[i]/A[j]])%MOD;
            }
            res = (res+m[A[i]])%MOD;
        }
        return res;
    }
};

File: 0823 - Binary Trees With Factors/solution.py
class Solution:
    def numFactoredBinaryTrees(self, A: List[int]) -> int:
        @lru_cache(None)
        def dp(n):
            return 1+sum(dp(i)*dp(n//i) for i in A if n%i == 0 and n//i in A)
        A = set(A)
        return sum(dp(i) for i in A)%int(1e9+7)